lib/model/folder_sendrecv_test.go

```diff
@@ -415,6 +416,87 @@ func TestWeakHash(t *testing.T) {
 	}
 }
 
+func TestWeakHashWrite(t *testing.T) {
+	// End-to-end test: if all blocks can be found by their weak hash,
+	// then the weak hash finder must be able to write a complete file.
+	const nblocks = 3
+	const size = nblocks * protocol.MinBlockSize
+
+	model, fo := setupSendReceiveFolder()
+	defer cleanupSRFolder(fo, model)
+	ffs := fo.Filesystem()
+
+	desired, err := ffs.Create("weakhash")
+	must(t, err)
+	defer desired.Close()
+	_, err = io.CopyN(desired, rand.Reader, size)
+	must(t, err)
+	_, err = desired.Seek(0, os.SEEK_SET)
+	must(t, err)
+
+	blocks, err := scanner.Blocks(context.TODO(), desired,
+		protocol.MinBlockSize, size, nil, true)
+	must(t, err)
+	if len(blocks) != nblocks {
+		t.Fatalf("#blocks = %d", len(blocks))
+	}
+
+	source, err := ffs.Create(fs.TempName("weakhash"))
+	must(t, err)
+	defer source.Close()
+
+	desiredInfo := protocol.FileInfo{
+		Name:   source.Name(),
+		Blocks: blocks,
+		Size:   size,
+	}
+
+	// Change the order of blocks from the desired file.
+	var buf [protocol.MinBlockSize]byte
+	for i := 0; i < nblocks; i++ {
+		shifted := (i + 1) % nblocks
+		_, err := desired.ReadAt(buf[:], int64(shifted*protocol.MinBlockSize))
+		must(t, err)
+		_, err = source.WriteAt(buf[:], int64(i*protocol.MinBlockSize))
```

Comment:
Out of curiosity it would be interesting to write the blocks not exactly at the block boundary, to verify that rolling actually finds it.

Reply:
Ah, nvm, it already does that.